src/mailmap.rs

```diff
@@ -0,0 +1,215 @@
+use crate::commit::Commit;
+use crate::repository::Repository;
+use crate::signature::{Signature, SignaturePayload};
+use napi::bindgen_prelude::*;
+use napi_derive::napi;
+use std::ops::{Deref, DerefMut};
+
+pub(crate) enum MailmapInner {
+  Repo(SharedReference<Repository, git2::Mailmap>),
+  Owned(git2::Mailmap),
+}
+
+impl Deref for MailmapInner {
+  type Target = git2::Mailmap;
+
+  fn deref(&self) -> &Self::Target {
+    match self {
+      Self::Repo(repo) => repo.deref(),
+      Self::Owned(mailmap) => mailmap,
+    }
+  }
+}
+
+impl DerefMut for MailmapInner {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    match self {
+      Self::Repo(repo) => repo.deref_mut(),
+      Self::Owned(mailmap) => mailmap,
+    }
+  }
+}
+
+#[napi(object)]
+pub struct AddMailmapEntryData {
+  pub real_name: Option<String>,
+  pub real_email: Option<String>,
+  pub replace_name: Option<String>,
+  pub replace_email: String,
+}
+
+/// A wrapper around git2::Mailmap providing Node.js bindings
+#[napi]
+pub struct Mailmap {
+  pub(crate) inner: MailmapInner,
+}
+
+#[napi]
+impl Mailmap {
+  /// Add a new Mailmap entry.
+  ///
+  /// Maps an author/committer (specified by `replace_name` and `replace_email`)
+  /// to the specified real name and email. The `replace_email` is required but
+  /// the other parameters can be null.
+  ///
+  /// If both `replace_name` and `replace_email` are provided, then the entry will
+  /// apply to those who match both. If only `replace_name` is provided,
+  /// it will apply to anyone with that name, regardless of email. If only
+  /// `replace_email` is provided, it will apply to anyone with that email,
+  /// regardless of name.
+  ///
+  /// @param {AddMailmapEntryData} entry - The mailmap entry data.
+  /// @returns {void}
+  /// @throws An error if the operation failed.
+  ///
+  /// @category Mailmap/Methods
+  ///
+  /// @signature
+  /// ```ts
+  /// class Mailmap {
+  ///   addEntry(entry: AddMailmapEntryData): void;
+  /// }
+  /// ```
+  #[napi]
+  pub fn add_entry(&mut self, entry: AddMailmapEntryData) -> crate::Result<()> {
+    self.inner.add_entry(
+      entry.real_name.as_deref(),
+      entry.real_email.as_deref(),
+      entry.replace_name.as_deref(),
+      &entry.replace_email,
+    )?;
+    Ok(())
+  }
+
+  /// Resolve a signature to its canonical form using a mailmap.
+  ///
+  /// Returns a new signature with the canonical name and email.
+  ///
+  /// @param {SignaturePayload} signature - Signature to resolve
+  /// @returns The resolved signature with canonical name and email
+  ///
+  /// @category Mailmap/Methods
+  ///
+  /// @signature
+  /// ```ts
+  /// class Mailmap {
+  ///   resolveSignature(signature: SignaturePayload): Signature;
+  /// }
+  /// ```
+  #[napi]
+  pub fn resolve_signature(&self, signature: SignaturePayload) -> crate::Result<Signature> {
+    let git_signature = git2::Signature::try_from(Signature::try_from(signature)?)?;
+
+    let resolved = match &self.inner {
```

Comment:
```suggestion
    let resolved = self.inner.resolve_signature(&git_signature)?;
```
Can we simplified here too?